webhook/resources/volumeattachment/validator.go

```diff
@@ -57,6 +60,15 @@ func (v *volumeAttachmentValidator) Update(request *admission.Request, oldObj ru
 	if !ok {
 		return werror.NewInvalidError(fmt.Sprintf("%v is not a *longhorn.VolumeAttachment", newObj), "")
 	}
+	isRemovingLonghornFinalizer, err := common.IsRemovingLonghornFinalizer(oldObj, newObj)
+	if err != nil {
+		err = errors.Wrap(err, "failed to check if removing longhorn.io finalizer from deleted object")
+		return werror.NewInvalidError(err.Error(), "")
+	} else if isRemovingLonghornFinalizer {
+		// We always allow the removal of the longhorn.io finalizer while an object is being deleted. It is the
+		// controller's responsibility to wait for the correct conditions to attempt to remove it.
+		return nil
+	}
```

Comment:
I am wondering the purpose of this code block. Looks like we check but always allow?

Reply:
The goal is to always allow the deletion of the Longhorn finalizer, regardless of any other rules in place. The context can be found in https://github.com/longhorn/longhorn/issues/7553 and https://github.com/longhorn/longhorn-manager/pull/2425. The basic idea is that sometimes:

- A controller may choose not to meaningfully continue to reconcile a resource after its deletionTimestamp is set.
- Instead, the controller wants to delete the resource.
- But some webhook check fails and disallows the deletion.

I did not test it, but I thought this may be a possible situation introduced by these changes.

- The volume attachment controller sees the volume is gone and the volume attachment is being deleted. It decids to remove the finalizer for the volume attachment. 
  https://github.com/longhorn/longhorn-manager/blob/ffe4a02e240cee2299967c38ab3ab37a6d13abf0/controller/volume_attachment_controller.go#L293-L302
- At this moment the new validation logic I wrote disallows updating the volume attachment.
- The situation never changes (since the volume attachment controller no longer meaningfully reconciles).

Since the goal in https://github.com/longhorn/longhorn/issues/7553 is to add this logic to all resources in the webhook eventually, I thought it was safest to add it here now while adding logic to avoid any issues. WDYT?

Reply:
Sounds good to me. Thank you for the explanation